development/source-code-documentation.md

```diff
@@ -0,0 +1,187 @@
+# Documentation standards
+
+Code should be documented according to [JSDoc](https://jsdoc.app/) standards. Documentation comments should enable the
+TypeScript compatible [type checking](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html),
+if such type checking is supported by an editor or an IDE. For this we should follow the 
+[TypeScript JSDoc guidelines](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html).
+
+Documentation should follow the [ESLint JSDoc rules](https://eslint.org/docs/rules/valid-jsdoc).
+
+## Module
+An ES6 module should include a module name at the first line of the module file:
+```
+/** @module moduleName */
+```
+
+A module name should start from a lowercase letter in order to distinguish it from the class name.
+
+## Enums
+
+Define enums as:
+```
+/**
+ * Enum description
+ *
+ * @enum {string} */
+export const EnumType = {
+  VALUE_ONE: 'one',
+  VALUE_TWO: 'two'
+}
+```
+
+References:
+[TypeScript enum documentation guidelines](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#enum)
+
+
+## Classes
+
+```
+class C {
+  /**
+   * @param {number} data
+   */
+  constructor(data) {
+    // Property types can be inferred
+    this.name = "foo"
+
+    // or set explicitly
+    /** @type {string | null} */
+    this.title = null
+
```

Comment:
@kirlat , do you mean that such comments should be inside methods to describe each class property?
Do you really like such approach? From my point of view it is difficult to update and support such code with so a lot of comments/ It is really difficult to read it - because it becomes much bigger. And it is a mix of comments and a real code.

And also why do we need such detailed information about types?  Could we place such information inside unit tests?

Reply:
I suggest to comment this way all the public props: because they're used outside of the class, it is important for those unfamiliar with the class internals to understand what those props are. We can also comment private props to remember, for ourselves, what those are too.

I do not think the it would be too hard to keep and update those comments because:
* We do not have too many props usually.
* Once the prop is created, it is hard to change it: since it already is used in many places, changing it would have last expanding consequences. So existing props, probably, would change rarely.
* We already have comment on methods in a way that is similar and we usually have less props than we have the methods. So probably it would be bearable.

Personally, I do not like it very much. It's too verbose to my taste and the TypeScript type definitions are much more concise. TypeScript, however, brings a lot of complexities of its own, and I don't think it would make sense to start using TypeScript in our project. But, on my, opinion, with all it's drawbacks, types through comments are still much better than nothing.

Type hinging and type checking are, on my opinion, great help to write a more reliable code. When refactoring the `Language` usages I had many complex situations where I had to untangle the allowed range of values through a chain of interdependent classes. I wish it would be described somewhere instead. One of the examples was the usage of `en-US` vs `eng`. There were some other situations like that.

Here are some of the screenshots of how it works in my IDE.

This is how hints of method parameters are displayed:
![image](https://user-images.githubusercontent.com/18631055/103349673-eddb2f00-4ab6-11eb-904f-965245f1831b.png)

Prop hinting in action. Once one hover over the prop usage code, the following pop-up appears:
![image](https://user-images.githubusercontent.com/18631055/103349789-427eaa00-4ab7-11eb-87f3-e5f06133e740.png)
Text in blue are hyperlinks and they could provide further information within the same pop-up.

Once the value of incorrect type is assigned to the prop, incorrect value is underlined:
![image](https://user-images.githubusercontent.com/18631055/103349916-b15c0300-4ab7-11eb-83bd-7ced1fd783ed.png)

Hovering over the prop in question triggers a pop-up with more details:
![image](https://user-images.githubusercontent.com/18631055/103349950-d9e3fd00-4ab7-11eb-8276-6d35db090f10.png)

Personally, I found those hints to be extremely important when working with (especially during refactoring) complex objects. I think this could increase productivity, and, probably more importantly, reliability of our code.

Regarding the size and readability, in my IDE comment sections could be collapsed:
![image](https://user-images.githubusercontent.com/18631055/103350209-a6ee3900-4ab8-11eb-896c-fa034d0fd3dd.png)
![image](https://user-images.githubusercontent.com/18631055/103350227-b53c5500-4ab8-11eb-8d6f-637aeac0944c.png)
But I don't know how is it in your case.

But that's just me. So I'm offering this for discussion and I'm very interested to hear your opinion about this. We can make it a rule if we decide to, but so far I'm offering this just as a suggestion. @balmas, @irina060981, what do you think?

Reply:
I tend to agree with @kirlat here. Wasn't it also the case that if we use such comments we could get the linter to warn us if we made incorrect assignments?

Reply:
Unfortunalely, VS code couldn't collapse/uncollapse comments - only methods.
I think that hints work well.
From my point of view - if anyone in the team says that such information could help him to refactor the code than I should do that :)
But as for me (as I could not collapse comments) - it would be more difficult to read the code. I use comments really very very rarely.
I think that there are only some useful comments about some code specifics, when it is decided to use not very straightforward solution.
As for properties - when I need to refactor I trace the code and go through the chains.
And I think that if we would define some standards for public,protected and private properties/methods - we would be able to reduce such chains.
I am not sure - what will take less time - trace the code or support comments.

Reply:
I think sometimes comments are important for explaining certain things that are not clear from the code. For example, why a particular approach was chosen over other ways of implementation. It can also be critical to understand business context that was behind the code; it's often impossible to get that from just reading the code itself.

There are different approaches to conveying that information: comments in the code itself, readme files, separate wikis, some teams even use Git comments. Some teams also has a zero-coment policy; I personally think this is a bit radical). None of those is perfect, on my opinion. So it's always a compromise. I think we should select our approach understanding all advantages and limitations of it.

I personally tend to think that when comments are together with the code there is less chance comments would be outdated. When something is changed, it is easier to update comment that is nearby right away to reflect the change rather than make changes to a separate file, which I would probably tend to delay. If comments are not up to date, it's easier to notice that during code reviews; I had several occasions of that noticed during my code reviews, and that was extremely helpful.

I agree that long comments could make code reading harder, However, in my case, they're shown in subdued colors in my IDE so visually they tend to blend with background and it's easier for reader to concentrate on the code. The comment blocks could also be collapsed in my case, as I already mentioned.

But despite of all that being said, I'm open to any effective way to store information that is not obvious from the code. I just think we should probably not match different approaches together because it means valuable information would be dispersed across several channels.

@irina060981, I'm not using VS Code myself, but I found some info about collapsing comments in it. Probably it might be possible to collapse comment blocks at least for some languages: https://www.youtube.com/watch?v=oFylGMaTCTs
> There are standard commands for that - just open the command palette and look for Fold All Block Comments and Unfold All (keybindings should be displayed along if available; if not, you can assign them yourself). Fold All Block Comments seems to work for multiple consecutive line comments (//) and C-style block comments (wrapped in /* and */) in TypeScript, YMMV for other languages (I assume this is dependent on each language's language server).

I'm wondering if it will work for you (that's regardless of the topic of whether long comments should be used or not).

I do support strongly the type information for properties comments, though. For me, at least, that would be a huge time saver and a way to reduce errors.

Reply:
@irina060981 and @kirlat here's what I think we agreed to re comments in our check-in today:

1. Method signatures and parameter types should be commented
2. "Why" comments should be present next to code which is unclear or which has business logic rationale that isn't obvious 
3. We should generally avoid public properties, which should make it less critical that we comment type information for them,  and when feasible, we should use validation methods inside the code to make sure properties can't be set to incorrect types.

Is there anything else we should clarify?

Reply:
I think everything is covered.

Reply:
Agree